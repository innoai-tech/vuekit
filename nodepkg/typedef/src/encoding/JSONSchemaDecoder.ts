import { type TypeAny } from "../Type";
import * as t from "../t";
import {
  assign,
  filter,
  has,
  isArray,
  isBoolean,
  map,
  some,
  last,
  mapValues,
  isPlainObject,
  isEmpty
} from "@innoai-tech/lodash";
import type { JSONSchema } from "./JSONSchemaEncoder";

export const refName = (ref: string) => {
  return last(ref.split("/"))!;
};

export class JSONSchemaDecoder {
  static decode(
    type: JSONSchema,
    resolveRef: (ref: string) => [JSONSchema, string]
  ): TypeAny {
    return new JSONSchemaDecoder(resolveRef).decode(type);
  }

  def = new Map<string, TypeAny>();

  constructor(private resolveRef: (ref: string) => [JSONSchema, string]) {
  }

  decode(jsonSchema: JSONSchema): TypeAny {
    const tt = this._decode(jsonSchema);
    if (jsonSchema && jsonSchema["description"]) {
      return tt.annotate({
        // only pick the first line as description
        description: jsonSchema["description"].split("\n")[0]
      });
    }
    return tt;
  }

  private _decode(schema: JSONSchema): TypeAny {
    schema = normalizeSchema(schema);

    if (schema["$ref"]) {
      const [reffedSchema, refName] = this.resolveRef(schema["$ref"]);

      if (!this.def.has(refName)) {
        this.def.set(refName, t.any());
        this.def.set(refName, this.decode(reffedSchema));
      }

      return t.ref(refName, () => this.def.get(refName)!);
    }

    if (schema["enum"]) {
      if (schema["enum"].length == 1) {
        return t.literal(schema["enum"][0]);
      }
      return t.enums(schema["enum"]);
    }

    if (schema["oneOf"]) {
      if (schema["discriminator"]) {
        const discriminatorPropertyName = schema["discriminator"][
          "propertyName"
          ] as string;
        if (discriminatorPropertyName) {
          const mapping: Record<string, any> = {};

          for (const o of schema["oneOf"]) {
            const tt = this.decode(o);

            if (tt) {
              const objectSchema: Record<string, any> = {};
              const values: any[] = [];

              for (const [propName, _, p] of tt.entries(
                {},
                { path: [], branch: [] }
              )) {
                if (p.type === "never") {
                  continue;
                }

                if (propName === discriminatorPropertyName) {
                  switch (p.type) {
                    case "literal": {
                      values.push(p.schema);
                      break;
                    }
                    case "enums": {
                      values.push(...Object.values(p.schema as Record<string, any>));
                    }
                  }
                  continue;
                }

                objectSchema[propName] = p;
              }

              if (values.length) {
                for (const value of values) {
                  mapping[value] = isEmpty(objectSchema) ? t.object() : t.object(objectSchema);
                }
              }
            }
          }

          return t.discriminatorMapping(discriminatorPropertyName, mapping);
        }
      }

      const oneOf = map(schema["oneOf"], (s) => this.decode(s));

      if (oneOf.length === 1) {
        return oneOf[0]!;
      }

      return t.union(...oneOf);
    }

    if (schema["allOf"]) {
      const allOf = map(schema["allOf"], (s) => this.decode(s));

      if (allOf.length === 1) {
        return allOf[0]!;
      }

      return t.intersection(...allOf);
    }

    if (isObjectType(schema)) {
      if (isPlainObject(schema["additionalProperties"])) {
        return t.record(
          this.decode(schema["propertyNames"] ?? { type: "string" }),
          this.decode(schema["additionalProperties"])
        );
      }

      const required = schema["required"] ?? [];

      return t.object(
        mapValues(schema["properties"], (s, n) => {
          const propType = this.decode(s);
          if (required.includes(n)) {
            return propType;
          }
          return propType.optional();
        })
      );
    }

    if (isArrayType(schema)) {
      if (isArray(schema["items"])) {
        return t.tuple(
          (schema["items"] as JSONSchema[]).map((s) => this.decode(s)) as any
        );
      }

      return t.array(this.decode(schema["items"]));
    }

    if (isStringType(schema)) {
      if (schema["format"] === "binary") {
        return t.binary();
      }
      return t.string();
    }

    if (isNumberType(schema)) {
      if (schema.type === "integer") {
        return t.integer();
      }
      return t.number();
    }

    if (isBooleanType(schema)) {
      return t.boolean();
    }

    if (isNullType(schema)) {
      return t.nil();
    }

    return t.any();
  }
}

const isObjectType = (schema: any): boolean => schema.type === "object";
const isNullType = (schema: any): boolean => schema.type === "null";
const isArrayType = (schema: any): boolean => schema.type === "array";
const isNumberType = (schema: any): boolean =>
  schema.type === "number" || schema.type === "integer";
const isStringType = (schema: any): boolean => schema.type === "string";
const isBooleanType = (schema: any): boolean => schema.type === "boolean";

const typeRelationKeywords: { [k: string]: string[] } = {
  object: [
    "properties",
    "additionalProperties",
    "unevaluatedProperties",
    "patternProperties",
    "propertyNames",
    "dependentSchemas",

    "maxProperties",
    "minProperties"
    // "required",
    // "dependentRequired",
  ],
  array: [
    "contains",
    "items",
    "additionalItems",
    "unevaluatedItems",

    "maxItems",
    "minItems",
    "uniqueItems",
    "maxContains",
    "minContains"
  ],
  string: [
    "pattern",
    "contentMediaType",
    "contentEncoding",
    "contentSchema",
    "maxLength",
    "minLength"
  ],
  number: [
    "maximum",
    "minimum",
    "multipleOf",
    "exclusiveMaximum",
    "exclusiveMinimum"
  ]
};

const hasProps = <T>(schema: T, props: Array<keyof T>): boolean =>
  some(props, (prop: string) => has(schema, prop));

const isMetaType = (schema: any): any => {
  return !hasProps(schema, ["type", "$ref", "$id", "oneOf", "anyOf", "allOf"]);
};

const normalizeSchema = (schema: any): any => {
  if (isBoolean(schema)) {
    return {};
  }

  // auto complete schema type
  if (!schema.type) {
    for (const t in typeRelationKeywords) {
      if (hasProps(schema, typeRelationKeywords[t] as any)) {
        schema.type = t as any;
        break;
      }
    }
  }

  if (schema.const) {
    schema.enum = [schema.const];
  }

  const mayNormalizeMeta = (key: "allOf" | "anyOf" | "oneOf") => {
    if (schema[key]) {
      schema[key] = filter(schema[key], (s) => {
        const ns = normalizeSchema(s);
        if (isMetaType(ns)) {
          if (key === "allOf") {
            // only allOf will merge meta
            assign(schema, ns);
          }
          return false;
        }
        return true;
      });

      if (schema[key]!.length === 0) {
        schema[key] = undefined;
      }
    }
  };

  mayNormalizeMeta("allOf");
  mayNormalizeMeta("anyOf");
  mayNormalizeMeta("oneOf");

  return schema;
};
