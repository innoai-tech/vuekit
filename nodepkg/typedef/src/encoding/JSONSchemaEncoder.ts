import { isType, Schema, type Type } from "../core";
import { isArray, isObjectLike } from "../core/util.ts";

export type JSONSchema = {
  type?: string;
  [x: string]: any;
};

export class JSONSchemaEncoder {
  static encode<T extends Type>(type: T): JSONSchema | false {
    return new JSONSchemaEncoder().encode(type);
  }

  def = new Map<string, JSONSchema | false>();

  encode<T extends Type>(type: T): JSONSchema | false {
    const s = this._encode(type);

    const $defs: Record<string, any> = {};

    for (const [name, d] of this.def) {
      $defs[name] = d;
    }

    return Object.assign(s, {
      $defs: $defs,
    });
  }

  private _encode<T extends Type>(type: T): JSONSchema | false {
    let schema = this._encodeSchema(type);

    if (Schema.metaProp(type, "title")) {
      schema = {
        ...schema,
        title: Schema.metaProp(type, "title"),
      };
    }

    if (Schema.metaProp(type, "description")) {
      schema = {
        ...schema,
        description: Schema.metaProp(type, "description"),
      };
    }

    return schema;
  }

  private _encodeSchema<T extends Type>(type: T): JSONSchema | false {
    if (Schema.schemaProp(type, "$ref")) {
      const refName = Schema.schemaProp(type, "$ref");

      if (refName) {
        if (!this.def.has(refName)) {
          // set to lock to avoid loop
          this.def.set(refName, {});

          const unwrap = (type as any).schema[Schema.unwrap];
          this.def.set(refName, this._encode(unwrap()));
        }

        return {
          $ref: `#/$defs/${refName}`,
        };
      }

      return this._encode(type);
    }

    return this._normalizeSchema(type.schema!);
  }

  private _normalizeSchema(s: Record<string, any> | null): JSONSchema | false {
    if (!s) {
      if (s === false) {
        return false;
      }
      return {};
    }

    const schema: Record<string, any> = {};

    for (const k of Object.getOwnPropertyNames(s)) {
      const v = s[k];

      if (isType(v)) {
        schema[k] = this._encode(v);
      } else if (isArray(v)) {
        schema[k] = v.map((item) => {
          if (isType(item)) {
            return this._encode(item);
          }
          return item;
        });
      } else if (isObjectLike(v)) {
        schema[k] = this._normalizeSchema(v);
      } else {
        schema[k] = v;
      }
    }

    return schema;
  }
}
