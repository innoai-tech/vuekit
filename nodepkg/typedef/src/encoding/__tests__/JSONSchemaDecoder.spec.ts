import { describe, expect, it } from "bun:test";
import { get } from "@innoai-tech/lodash";
import { JSONSchemaDecoder, refName } from "../JSONSchemaDecoder.ts";
import { JSONSchemaEncoder } from "../JSONSchemaEncoder.ts";

describe("JSONSchemaDecoder", () => {
  it("decode complex", () => {
    const schema = {
      $defs: {
        A: { type: "string" },
        B: { type: "integer" },
        C: { type: "string", enum: ["X", "Y", "Z"] },
        ObjC: {
          type: "object",
          properties: {
            c: { $ref: "#/$defs/C" },
          },
        },
        Obj: {
          allOf: [
            { $ref: "#/$defs/ObjC" },
            {
              type: "object",
              properties: {
                a: { $ref: "#/$defs/A" },
                nested: { $ref: "#/$defs/ObjC" },
              },
              required: ["a"],
            },
          ],
        },
        Arr: {
          type: "array",
          items: { type: "string" },
        },
        Map: {
          type: "object",
          additionalProperties: {
            type: "string",
          },
        },
        Union: {
          type: "object",
          discriminator: {
            propertyName: "type",
          },
          required: ["type"],
          oneOf: [
            {
              properties: {
                type: { enum: ["A"] },
                a: { $ref: "#/$defs/A" },
                b: { $ref: "#/$defs/B" },
              },
              required: ["type", "a"],
              additionalProperties: false,
            },
            {
              properties: {
                type: { enum: ["B"], type: "string" },
                b: { $ref: "#/$defs/B" },
              },
              required: ["type", "b"],
              additionalProperties: false,
            },
          ],
        },
        TaggedUnionWhenA: {
          properties: {
            type: { enum: ["A"] },
            a: { $ref: "#/$defs/A" },
            b: { $ref: "#/$defs/B" },
          },
          required: ["type", "a"],
          additionalProperties: false,
        },
        TaggedUnion: {
          type: "object",
          discriminator: {
            propertyName: "type",
            mapping: {
              A: { $ref: "#/$defs/TaggedUnionWhenA" },
              B: {
                properties: {
                  type: { enum: ["B"], type: "string" },
                  b: { $ref: "#/$defs/B" },
                },
                required: ["type", "b"],
                additionalProperties: false,
              },
            },
          },
        },
      },
      type: "object",
      additionalProperties: false,
      properties: {
        obj: { $ref: "#/$defs/Obj" },
        arr: { $ref: "#/$defs/Arr" },
        map: { $ref: "#/$defs/Map" },
        union: { $ref: "#/$defs/Union" },
        tagged: { $ref: "#/$defs/TaggedUnion" },
      },
    };

    const t = JSONSchemaDecoder.decode(schema, (ref) => {
      return [get(schema, ref.slice(2).split("/"), {}), refName(ref)];
    });

    const schemaV2 = JSONSchemaEncoder.encode(t);
    expect(schemaV2).toMatchSnapshot();
  });
});
