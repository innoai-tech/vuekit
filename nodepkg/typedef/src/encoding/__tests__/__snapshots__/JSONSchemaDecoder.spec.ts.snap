// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaDecoder decode complex 1`] = `
{
  "$defs": {
    "A": {
      "type": "string",
    },
    "Arr": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
    "B": {
      "type": "integer",
    },
    "C": {
      "enum": [
        "X",
        "Y",
        "Z",
      ],
    },
    "Map": {
      "additionalProperties": {
        "type": "string",
      },
      "propertyNames": {
        "type": "string",
      },
      "type": "object",
    },
    "Obj": {
      "allOf": [
        {
          "$ref": "#/$defs/ObjC",
        },
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "$ref": "#/$defs/A",
            },
            "nested": {
              "$ref": "#/$defs/ObjC",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
      ],
    },
    "ObjC": {
      "additionalProperties": false,
      "properties": {
        "c": {
          "$ref": "#/$defs/C",
        },
      },
      "required": [],
      "type": "object",
    },
    "TaggedUnion": {
      "discriminator": {
        "propertyName": "type",
      },
      "oneOf": [
        {
          "$ref": "#/$defs/TaggedUnionWhenA",
        },
        {
          "additionalProperties": false,
          "properties": {
            "b": {
              "$ref": "#/$defs/B",
            },
            "type": {
              "enum": [
                "B",
              ],
            },
          },
          "required": [
            "type",
            "b",
          ],
          "type": "object",
        },
      ],
    },
    "TaggedUnionWhenA": {
      "additionalProperties": false,
      "properties": {
        "a": {
          "$ref": "#/$defs/A",
        },
        "b": {
          "$ref": "#/$defs/B",
        },
        "type": {
          "enum": [
            "A",
          ],
        },
      },
      "required": [
        "type",
        "a",
      ],
      "type": "object",
    },
    "Union": {
      "discriminator": {
        "propertyName": "type",
      },
      "oneOf": [
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "$ref": "#/$defs/A",
            },
            "b": {
              "$ref": "#/$defs/B",
            },
            "type": {
              "enum": [
                "A",
              ],
            },
          },
          "required": [
            "type",
            "a",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "b": {
              "$ref": "#/$defs/B",
            },
            "type": {
              "enum": [
                "B",
              ],
            },
          },
          "required": [
            "type",
            "b",
          ],
          "type": "object",
        },
      ],
    },
  },
  "additionalProperties": false,
  "properties": {
    "arr": {
      "$ref": "#/$defs/Arr",
    },
    "map": {
      "$ref": "#/$defs/Map",
    },
    "obj": {
      "$ref": "#/$defs/Obj",
    },
    "tagged": {
      "$ref": "#/$defs/TaggedUnion",
    },
    "union": {
      "$ref": "#/$defs/Union",
    },
  },
  "required": [],
  "type": "object",
}
`;
